EVS global_ens duplicate station and related bugfixes

StevenSimon-NOAA opened the pull request:
## Pull Request Testing ##

- [ ] Describe testing already performed for this Pull Request:</br>

1. Introduced duplicate flag into metplus override option within local global_ens wave config file and generated a g2o wave stats file with MetPlus 5.0.1 that matches the one generated by the MetPlus 6.0 Beta in emc.vpppg EVS Parallel.

2. Did local runs of all global_ens stats drivers to confirm that the CLIMO_FILE_NAME option is suitable to supplant the CLIMO_INPUT_DIR and CLIMO_INPUT_TEMPLATE options in global_ens config files that would trigger deprecation warnings in the emc.vpppg EVS Parallel MetPlus 6.0 Beta. Repeated similar process of testing with the update of PB2NC_SKIP_TIMES to PB2NC_SKIP_VALID_TIMES in parm/metplus_config/prep/global_ens/wave_grid2obs/PB2NC_wave.conf, but with local runs of the global_ens wave prep driver. Not applicable for the strict purposes of this PR.   

3. Confirmed that switching FCST_PCP_COMBINE_USE_ZERO_ACCUM back to False in the ecme 24h snowfall stats config file does not impact the associated stats generated.   

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

1. Compare global_ens g2o (atmos and wave) stats files with duplicate flag bugfix against those generated by emc.vpppg EVS Parallel MetPlus 6.0 Beta. Confirm that the duplicate stations are indeed removed, i.e., smaller stats file sizes.

2. Run EVS global_ens stats jobs and global_ens wave prep job with the MetPlus 6.0 Beta and confirm that all deprecation warnings have been properly handled, i.e., they no longer exist in the associated logfiles. Not applicable for the strict purposes of this PR.

3. Verify that switching FCST_PCP_COMBINE_USE_ZERO_ACCUM back to False in the ecme 24h snowfall stats config file does not impact the associated stats generated. This would involve a  comparison of the ecme snowfall stats files generated locally, by the emc.vpppg MetPlus 6.0 Beta, and by nco prod. The ecme snowfall stats files generated should be the exact same.       

- [ ] Has the code been checked to ensure that no errors occur during the execution? **[Yes or No]**

Tentatively yes

- [ ] Do these updates/additions include sufficient testing updates? **[Yes or No]**

Yes most likely

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

ASAP

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

StevenSimon-NOAA committed:
Align global_ens climo field dictionaries with 10 or greater variables with the alaphabetical alignment of the forecast and obs variable dictionaries.
This convention is solely for MetPlus 5.0.1 as a placeholder bugfix.

StevenSimon-NOAA committed:
Update deprecated PB2NC_SKIP_TIMES and implement duplicate flag in temporary override as a bugfix.

StevenSimon-NOAA committed:
Implement duplicate flag in override as a temporary bugfix in global_ens wave.

StevenSimon-NOAA committed:
Make updates in global_ens stats config files to address GRID_STAT and POINT_STAT deprecation WARNINGs.

StevenSimon-NOAA committed:
Update duplicate flag override for global_ens wave.

StevenSimon-NOAA committed:
Revert climo field ordering schema. The MetPlus 5.0.1 alphabetical climo field ordering schema bugfix  will be applied in a separate feature branch and PR.

StevenSimon-NOAA committed:
Merge branch 'develop' into feature/global_ens_metplus_climo_config_bugfix. Image names were updated.

StevenSimon-NOAA committed:
Merge branch 'develop' into feature/global_ens_metplus_climo_config_bugfix. Update PcpCombine_fcstMESOSCALE_snow_vars.conf.

StevenSimon-NOAA committed:
Remove unnecessary commented out lines.

StevenSimon-NOAA committed:
Merge branch 'develop' into feature/global_ens_metplus_climo_config_bugfix. Implement Feature/2023tc.

StevenSimon-NOAA committed:
Set FCST_PCP_COMBINE_USE_ZERO_ACCUM=False for ecme 24h accum snowfall pcpcombine.

StevenSimon-NOAA committed:
Remove unnecessary commented out line in ecme 24h snowfall pcpcombine config file.

StevenSimon-NOAA committed:
Remove more unneccessary commented out lines and restore proper spacing.

AliciaBentley-NOAA commented:
@StevenSimon-NOAA Thank you for submitting this bugfix PR! I've updated the [List of PRs since EVS v1.0.8](https://docs.google.com/document/d/1h6cTQvooLpdX8zsUV8S7JFKPNAoNc54gPgq-fQ3DG40/edit?usp=sharing) document to include it. 

I think that your bugfix PR looks great, but I have one quick question for you: 
1) Are the changes made in [parm/metplus_config/prep/global_ens/wave_grid2obs/PB2NC_wave.conf](https://github.com/NOAA-EMC/EVS/pull/467/files#diff-1df3a7b66a18d9b3a6d3b2e02890f2932a84a4f4748835c5c45d327739fe05fb) to "PB2NC_SKIP_VALID_TIMES =" outside the scope of this bugfix PR? Can you remind me what is broken in ops EVS v1.0 that requires this change to be made?

StevenSimon-NOAA commented:
> @StevenSimon-NOAA Thank you for submitting this bugfix PR! I've updated the [List of PRs since EVS v1.0.8](https://docs.google.com/document/d/1h6cTQvooLpdX8zsUV8S7JFKPNAoNc54gPgq-fQ3DG40/edit?usp=sharing) document to include it.
> 
> I think that your bugfix PR looks great, but I have one quick question for you:
> 
> 1. Are the changes made in [parm/metplus_config/prep/global_ens/wave_grid2obs/PB2NC_wave.conf](https://github.com/NOAA-EMC/EVS/pull/467/files#diff-1df3a7b66a18d9b3a6d3b2e02890f2932a84a4f4748835c5c45d327739fe05fb) to "PB2NC_SKIP_VALID_TIMES =" outside the scope of this bugfix PR? Can you remind me what is broken in ops EVS v1.0 that requires this change to be made?

That is associated with another deprecation warning that appears in log files associated with MetPlus 6.0 Beta Processing. I forgot to list that explicitly in the PR.

AliciaBentley-NOAA commented:
@StevenSimon-NOAA Thanks! If this is not something that is broken in ops EVS v1.0 (which use MET-11.0.2/METplus-5.0.1) and will only arise when we switch to MET-12.0.0/METplus-6.0.0, then that change should be removed from this PR. We can not include **anything** in the EVS bugfix release that addresses issues that will start to occur when we switch to MET-12.0.0/METplus-6.0.0 in EVS v2.0. Please let me know if that makes sense, and also ensure that you are only addressing things that are **currently broken** in ops EVS v1.0. Thanks!

CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

StevenSimon-NOAA commented:
> @StevenSimon-NOAA Thanks! If this is not something that is broken in ops EVS v1.0 (which use MET-11.0.2/METplus-5.0.1) and will only arise when we switch to MET-12.0.0/METplus-6.0.0, then that change should be removed from this PR. We can not include **anything** in the EVS bugfix release that addresses issues that will start to occur when we switch to MET-12.0.0/METplus-6.0.0 in EVS v2.0. Please let me know if that makes sense, and also ensure that you are only addressing things that are **currently broken** in ops EVS v1.0. Thanks!
> 
> CC @PerryShafran-NOAA @ShelleyMelchior-NOAA

This does not make sense because this PR is set to be a develop-to-develop PR. Also, this will involve a lot of time to revert manually. Right now, we are not collectively on the same page, and I do not feel comfortable committing to undoing that many edits, only to have to possibly redo them.

StevenSimon-NOAA commented:
I am getting conflicting guidance here, and before moving forward, I would like to establish a common understanding for this PR.

AliciaBentley-NOAA commented:
@StevenSimon-NOAA Backstory: Perry and Shelley have been pushing **entire** PRs that have been merged into develop into the [EVS-1.0.x-bugfix branch](https://github.com/NOAA-EMC/EVS/tree/release/evs.v1.0.x-bugfix) thus far. They are using the [list of develop commits](https://github.com/NOAA-EMC/EVS/commits/develop/) in order to do this. If your current PR contains more than the EVS v1.0 bugfixes that should be added to the EVS-1.0.x-bugfix branch, Perry/Shelley will not be able to add your PR to the EVS-1.0.x-bugfix branch very easily. Perhaps @ShelleyMelchior-NOAA has told you that your PRs can contain more than what is currently broken in EVS v1.0 and that she will manually move individual files that you list for her into the EVS-1.0.x-bugfix branch. I am not aware of that discussion, if it has happened.

If your current bugfix PR could contain only the bugfixes that address what is currently broken in EVS v1.0, it could be added to the EVS-1.0.x-bugfix branch (in its entirety) much more easily. I hope that makes more sense now. Please work with @PerryShafran-NOAA and @ShelleyMelchior-NOAA to get this PR in a form that they can add push from develop into the EVS-1.0.x-bugfix branch.

StevenSimon-NOAA committed:
Merge branch 'develop' into feature/global_ens_metplus_climo_config_bugfix. Implement Feature/subseasonal HGT and precip prep.

StevenSimon-NOAA committed:
Revert MetPlus 6.0  Beta Deprecation Warning Bug Fixes.

StevenSimon-NOAA commented:
Ok, I reverted the MetPlus 6.0 Beta Deprecation Warning Bug Fixes.

StevenSimon-NOAA commented:
Standby for spacing issues I need to clean up.

StevenSimon-NOAA committed:
Fix additional spacing issues following the prior revert changes.

StevenSimon-NOAA committed:
Fix additional spacing issue in naefs headline config file.

StevenSimon-NOAA committed:
Fix "spacing issues" in config files.

ShelleyMelchior-NOAA commented:
This PR is in great shape now. Thank you, Steve.

PerryShafran-NOAA commented:
@StevenSimon-NOAA I'll be working on this PR.  Just so you know, I will be testing this PR with both the current prod versions of MET and METplus, and also the beta3 versions.  We'll take this PR one step at a time, and I'll be starting with the current prod versions.  This will go into the bugfix parallel.  Later, once that's all good and validated, I'll do a second test with the beta3 versions.  This will go into the main parallel.  

@AliciaBentley-NOAA and @ShelleyMelchior-NOAA for awareness.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA I'll be working on this PR. Just so you know, I will be testing this PR with both the current prod versions of MET and METplus, and also the beta3 versions. We'll take this PR one step at a time, and I'll be starting with the current prod versions. This will go into the bugfix parallel. Later, once that's all good and validated, I'll do a second test with the beta3 versions. This will go into the main parallel.
> 
> @AliciaBentley-NOAA and @ShelleyMelchior-NOAA for awareness.

As an update, this PR strictly addresses the duplicate flag issue and the ecme 24h snowfall stats config file reversion. Any handling of MetPlus 6.0 deprecation warnings was removed yesterday. I updated the PR accordingly.

PerryShafran-NOAA commented:
@StevenSimon-NOAA For test 1:

Currently, most of the jobs are done, with the gefs atmos grid2obs jobs continuing to run.  This is the request to run the global_ens grid2obs jobs, including wave, using the current production MET/METplus versions (11.0.2 and 5.0.1).  

The .o files are here:

/lfs/h2/emc/vpppg/noscrub/perry.shafran/pr467test/EVS/dev/drivers/scripts/stats/global_ens

The various stats output is here:  

/lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/global_ens

Please examine whether these stats jobs look good to you.  Again, as of this writing, the GEFS job is still running, but the rest of the jobs are complete. You may start to examine the completed jobs, and then check out the GEFS when complete (that should be soon because the job is 1:43 into a 2 hr job).  

Once we have validated these jobs, then we will repeat this test with the beta3 runs.  We need to ensure that they also work as they are expected with the beta3 versions of MET/METplus.

PerryShafran-NOAA commented:
BTW, the GEFS run is over, making all the runs complete for this first test.

StevenSimon-NOAA commented:
> /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/global_ens

The generated logs do not reveal any errors or WARNINGs. The small stats files do not match the full expected quantities for gefs, cmce, or ecme. That is because the emc.vpppg archive does not have 16 days of atmos prep data. They only have 15. This does not impact ecme small stats files counts because they only go out to F360. And the wave prep archive is sufficiently large to generate the full allotment of expected small wave stats files. 

With that said, the cmce, ecme, gefs (atmos and wave), and naefs large stats files were successfully generated.

PerryShafran-NOAA commented:
@StevenSimon-NOAA I just want to check - did you compare these to the emc.vpppg stats files?  You wanted to check to see if the file sizes were smaller, which successfully removed the duplicate stations compared to the emc.vpppg files.  Please check that out and let me know if this aspect of your PR was successful.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA I just want to check - did you compare these to the emc.vpppg stats files? You wanted to check to see if the file sizes were smaller, which successfully removed the duplicate stations compared to the emc.vpppg files. Please check that out and let me know if this aspect of your PR was successful.

In every atmos case but the ecme and naefs g2o stats, the test version of the large stats file was smaller than the emc.vpppg version. It is not clear to me why that is the case for ecme and naefs g2o stats. Nothing in the g2o config files jumps off the page as the reason for this behavior.  

emc.vpppg vpppg 37364978 Apr 18 11:33 evs.stats.ecme.atmos.grid2obs.v20240416.stat
perry.shafran vpppg 37365616 Apr 18 18:15 evs.stats.ecme.atmos.grid2obs.v20240416.stat

emc.vpppg vpppg 15718881 Apr 18 10:42 evs.stats.naefs.atmos.grid2obs.v20240416.stat
perry.shafran vpppg 15766640 Apr 18 17:19 evs.stats.naefs.atmos.grid2obs.v20240416.stat

Keep in mind these are the only new lines added to each g2o config file (EnsembleStat and PointStat): 

ENSEMBLE_STAT_MET_CONFIG_OVERRIDES = duplicate_flag = UNIQUE; obs_summary = NEAREST;
POINT_STAT_MET_CONFIG_OVERRIDES = duplicate_flag = UNIQUE; obs_summary = NEAREST;



Since global_ens wave already has the duplicate flag option in both versions, the large wave stats are exactly equivalent.

PerryShafran-NOAA commented:
@StevenSimon-NOAA OK, so the test stats file should actually be smaller than emc.vpppg, but in this case they are bigger and that is not correct?

StevenSimon-NOAA commented:
> @StevenSimon-NOAA OK, so the test stats file should actually be smaller than emc.vpppg, but in this case they are bigger and that is not correct?

That is my understanding, based on earlier discussions with DTC. And that correctly applies to the gefs and cmce stats files. But the ecme and naefs atmos test stats files are actually larger than their emc.vpppg counterparts. So a mixed bag. What does not help is that the atmos g2o config files did not have the duplicate flag option in them prior. The wave g2o stats config file simply did not have the duplicate flag option in an override option.

PerryShafran-NOAA commented:
Huh, I'm not sure what to suggest.  You might need to do a deep dive into each file to see exactly what's different, and why it might be larger.  Don't just do a mere diff because every line is different due to the different version numbers in each file.  Put the two files up side-by-side and see what might be different between the two that would explain why the test file is larger.

StevenSimon-NOAA commented:
> Huh, I'm not sure what to suggest. You might need to do a deep dive into each file to see exactly what's different, and why it might be larger. Don't just do a mere diff because every line is different due to the different version numbers in each file. Put the two files up side-by-side and see what might be different between the two that would explain why the test file is larger.

That would be going through tens of thousands of lines within each stats file. This is not something that is quickly or easily eyeballed.

PerryShafran-NOAA commented:
@StevenSimon-NOAA OK, here's what we're going to do:

I'm going to run in cron these jobs at the exact same time that the emc.vpppg job runs.  It could be that there might be more input in my run, which is later, which might be causing the job to be larger.  This might be a good test.  We'll have to get back to this tomorrow, then.  

@AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness

StevenSimon-NOAA commented:
> @StevenSimon-NOAA OK, here's what we're going to do:
> 
> I'm going to run in cron these jobs at the exact same time that the emc.vpppg job runs. It could be that there might be more input in my run, which is later, which might be causing the job to be larger. This might be a good test. We'll have to get back to this tomorrow, then.
> 
> @AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness

Sounds like a plan! I will repeat the same comparison for tomorrow's runs.

PerryShafran-NOAA commented:
@StevenSimon-NOAA The global_ens g2o atmos runs were completed at the same time as the emc.vpppg runs.  Please take a look at the output in the same directories and compare the filesizes.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA The global_ens g2o atmos runs were completed at the same time as the emc.vpppg runs. Please take a look at the output in the same directories and compare the filesizes.

Same issue as yesterday. The test ecme and naefs stats files are larger than their emc.vpppg counterparts. I reviewed the ecme and naefs g2o config files and there is nothing in them that differs from the cmce and gefs g2o config files. At this point, I do not not know what to look for in these stats files.

PerryShafran-NOAA commented:
@StevenSimon-NOAA I checked it out myself just now.  Here is what I see:

for emc.vpppg:
```
16 (dlogin01) /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/stats/global_ens > ls -l *20240417*/*grid2obs*
-rw-r--r-- 1 emc.vpppg vpppg  50845628 Apr 19 11:03 cmce.20240417/evs.stats.cmce.atmos.grid2obs.v20240417.stat
-rw-r--r-- 1 emc.vpppg vpppg  38239547 Apr 19 11:27 ecme.20240417/evs.stats.ecme.atmos.grid2obs.v20240417.stat
-rw-r--r-- 1 emc.vpppg vpppg 234501404 Apr 19 11:38 gefs.20240417/evs.stats.gefs.atmos.grid2obs.v20240417.stat
-rw-r--r-- 1 emc.vpppg vpppg   5261221 Apr 18 10:05 gefs.20240417/evs.stats.gefs.wave.grid2obs.v20240417.stat
-rw-r--r-- 1 emc.vpppg vpppg  16433307 Apr 19 10:38 naefs.20240417/evs.stats.naefs.atmos.grid2obs.v20240417.stat
```

For the test:

```
6 (dlogin07) /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/global_ens > ls -l *20240417*/*grid2obs*
-rw-r--r-- 1 perry.shafran vpppg  50860384 Apr 19 11:08 cmce.20240417/evs.stats.cmce.atmos.grid2obs.v20240417.stat
-rw-r--r-- 1 perry.shafran vpppg  38240387 Apr 19 11:33 ecme.20240417/evs.stats.ecme.atmos.grid2obs.v20240417.stat
-rw-r--r-- 1 perry.shafran vpppg 234221240 Apr 19 11:53 gefs.20240417/evs.stats.gefs.atmos.grid2obs.v20240417.stat
-rw-r--r-- 1 perry.shafran vpppg   5261221 Apr 18 16:43 gefs.20240417/evs.stats.gefs.wave.grid2obs.v20240417.stat
-rw-r--r-- 1 perry.shafran vpppg  16430317 Apr 19 10:38 naefs.20240417/evs.stats.naefs.atmos.grid2obs.v20240417.stat
```

Meaning for runs at the same time, emc.vpppg is larger naefs and for gefs, and emc.vpppg is smaller for cmce and ecme.  Which - is a different result than the previous test?  I'm not entirely sure what to make of this one.

@AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness and thoughts.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA I checked it out myself just now. Here is what I see:
> 
> for emc.vpppg:
> 
> ```
> 16 (dlogin01) /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/stats/global_ens > ls -l *20240417*/*grid2obs*
> -rw-r--r-- 1 emc.vpppg vpppg  50845628 Apr 19 11:03 cmce.20240417/evs.stats.cmce.atmos.grid2obs.v20240417.stat
> -rw-r--r-- 1 emc.vpppg vpppg  38239547 Apr 19 11:27 ecme.20240417/evs.stats.ecme.atmos.grid2obs.v20240417.stat
> -rw-r--r-- 1 emc.vpppg vpppg 234501404 Apr 19 11:38 gefs.20240417/evs.stats.gefs.atmos.grid2obs.v20240417.stat
> -rw-r--r-- 1 emc.vpppg vpppg   5261221 Apr 18 10:05 gefs.20240417/evs.stats.gefs.wave.grid2obs.v20240417.stat
> -rw-r--r-- 1 emc.vpppg vpppg  16433307 Apr 19 10:38 naefs.20240417/evs.stats.naefs.atmos.grid2obs.v20240417.stat
> ```
> 
> For the test:
> 
> ```
> 6 (dlogin07) /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/global_ens > ls -l *20240417*/*grid2obs*
> -rw-r--r-- 1 perry.shafran vpppg  50860384 Apr 19 11:08 cmce.20240417/evs.stats.cmce.atmos.grid2obs.v20240417.stat
> -rw-r--r-- 1 perry.shafran vpppg  38240387 Apr 19 11:33 ecme.20240417/evs.stats.ecme.atmos.grid2obs.v20240417.stat
> -rw-r--r-- 1 perry.shafran vpppg 234221240 Apr 19 11:53 gefs.20240417/evs.stats.gefs.atmos.grid2obs.v20240417.stat
> -rw-r--r-- 1 perry.shafran vpppg   5261221 Apr 18 16:43 gefs.20240417/evs.stats.gefs.wave.grid2obs.v20240417.stat
> -rw-r--r-- 1 perry.shafran vpppg  16430317 Apr 19 10:38 naefs.20240417/evs.stats.naefs.atmos.grid2obs.v20240417.stat
> ```
> 
> Meaning for runs at the same time, emc.vpppg is larger naefs and for gefs, and emc.vpppg is smaller for cmce and ecme. Which - is a different result than the previous test? I'm not entirely sure what to make of this one.
> 
> @AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness and thoughts.

That is definitely a different result. I reviewed the log files and confirmed that the duplicate flag and obs_summary options are read in correctly. The obs_summary option is set to NEAREST, in accordance with the global_det g2o (PointStat) config files. Bear in mind that the non-test global_ens atmos g2o config files do not have the duplicate flag or obs_summary options in them.

AliciaBentley-NOAA commented:
@StevenSimon-NOAA @PerryShafran-NOAA @ShelleyMelchior-NOAA A few questions and thoughts for you:

1) Steve- During your testing, were you able to successfully produce smaller stat files for gefs, cmce, naefs, and ecme? So you've proven that this works, it just isn't working in Perry's testing?

2) Is the version number of MET/METplus playing a role here? What versions are we testing with? Can we test with the ops versions first (MET-11.0.2/METplus-5.0.1) since this is an ops bugfix and go from there?

3) I'm not sure if this would cause the issues we're seeing or not, but I've noticed that different .conf files have different lines added. Some files have a long line like this:

`ENSEMBLE_STAT_MET_CONFIG_OVERRIDES = tmp_dir = "{TMP_DIR}"; duplicate_flag = UNIQUE; obs_summary = NEAREST;`

Other files have a shorter line, like this (nothing about tmp_dir):

`ENSEMBLE_STAT_MET_CONFIG_OVERRIDES = duplicate_flag = UNIQUE; obs_summary = NEAREST;`

Could this be making a difference in this PR? Why do these lines differ across .conf files?

AliciaBentley-NOAA commented:
@StevenSimon-NOAA @PerryShafran-NOAA @ShelleyMelchior-NOAA Also, please compare the results of your testing to the ops EVS stat files from NCO, as those are the stat files we're trying to make smaller by not duplicating stations. Thanks!

PerryShafran-NOAA commented:
@StevenSimon-NOAA @AliciaBentley-NOAA @ShelleyMelchior-NOAA I made a comparison of the test I just made to the ops stats files of today.  It's the same result.  In fact, the ops stats files pretty much match the emc.vpppg files.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA @PerryShafran-NOAA @ShelleyMelchior-NOAA A few questions and thoughts for you:
> 
> 1. Steve- During your testing, were you able to successfully produce smaller stat files for gefs, cmce, naefs, and ecme? So you've proven that this works, it just isn't working in Perry's testing?
> 2. Is the version number of MET/METplus playing a role here? What versions are we testing with? Can we test with the ops versions first (MET-11.0.2/METplus-5.0.1) since this is an ops bugfix and go from there?
> 3. I'm not sure if this would cause the issues we're seeing or not, but I've noticed that different .conf files have different lines added. Some files have a long line like this:
> 
> `ENSEMBLE_STAT_MET_CONFIG_OVERRIDES = tmp_dir = "{TMP_DIR}"; duplicate_flag = UNIQUE; obs_summary = NEAREST;`
> 
> Other files have a shorter line, like this (nothing about tmp_dir):
> 
> `ENSEMBLE_STAT_MET_CONFIG_OVERRIDES = duplicate_flag = UNIQUE; obs_summary = NEAREST;`
> 
> Could this be making a difference in this PR? Why do these lines differ across .conf files?

1. I did not have a chance do my own testing prior to this. Since we have a devprod switch occurring next week, it seemed that we were on an expedited schedule. 

2. I believe we are conducting our tests using apples-to-apples versions of Met and MetPlus.

3. I did not personally develop any of the global_ens config files. In the global_ens g2o config files without override options, I added them, i.e., ENSEMBLE_STAT_MET_CONFIG_OVERRIDES = duplicate_flag = UNIQUE; obs_summary = NEAREST;

There were several instances in the global_ens g2o config files where ENSEMBLE_STAT_MET_CONFIG_OVERRIDES = tmp_dir = "{TMP_DIR}";. In those cases, I appended the duplicate_flag and obs_summary options using the correct syntax. I do not know why some had this tmp_dir override and why some did not. I also do not know why the global_ens g2o config files did not have the duplicate flag and obs_summary options in the first place. Those were choices made by the prior developers that were not conveyed in the config files as comments or conveyed to me directly.

PerryShafran-NOAA commented:
@StevenSimon-NOAA All PRs should be fully tested and confirmed that they work before they are submitted for PR.  Having a machine switch upcoming shouldn't change that; I can test on any machine, and you should be able to do so as well.  

I need to suspend this PR so you can do some further testing yourself on these changes.  Please do some independent testing to be sure that your code fixes work as intended.  If there is confusion, please submit a question to the DTC discussion page, which is here:    https://github.com/dtcenter/METplus/discussions .  They are there to help us.  

@AliciaBentley-NOAA @ShelleyMelchior-NOAA

StevenSimon-NOAA commented:
> @StevenSimon-NOAA All PRs should be fully tested and confirmed that they work before they are submitted for PR. Having a machine switch upcoming shouldn't change that; I can test on any machine, and you should be able to do so as well.
> 
> I need to suspend this PR so you can do some further testing yourself on these changes. Please do some independent testing to be sure that your code fixes work as intended. If there is confusion, please submit a question to the DTC discussion page, which is here: https://github.com/dtcenter/METplus/discussions . They are there to help us.
> 
> @AliciaBentley-NOAA @ShelleyMelchior-NOAA

I will open an email thread with DTC. But last time we worked with them, they requested "troubleshooting aid" changes to the global_ens config files that caused every stats job to time out indefinitely. And sometimes they seem to prefer not to review our stats files because they do not like the formatting.

StevenSimon-NOAA commented:
Ok, DTC just responded to our query. It appears that the duplicate_flag and obs_summary modifications we made resulted in the expected stats processing. And they tend to think that the tmp_dir option in the metplus config override is not necessary. However, per Binbin's email response, the tmp_dir option in the metplus config override was added to certain g2o stats config files to deal with a past MetPlus bug in "an older version" of MetPlus. However, he does not recall whether this was addressed down the line in a newer version of MetPlus. 

With that in mind, comparing stats file sizes of the test and control cases may not be determinative enough for the purposes of our PR.

PerryShafran-NOAA commented:
@StevenSimon-NOAA OK, thanks for asking DTC!  Looks like we can go forward with at least this portion of this PR.  It seems to suggest that DTC says that all the expected changes in the stats files are correct.  

Can we check with DTC about Binbin's email, just to get confidence that the tmp_dir option is not necessary?   It seems that they lean towards not needing it, but they might need to know what Binbin's email had said so everyone can know whether this is actually addressed by METplus.

Finally, if you haven't done it yet, please be sure you have fully tested numbers 2 and 3 of your list of tests above in the instructions.  I want to be sure that you see them working before I do any PR testing on them.   When they are tested, please let me know.  

Thanks!

@AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA OK, thanks for asking DTC! Looks like we can go forward with at least this portion of this PR. It seems to suggest that DTC says that all the expected changes in the stats files are correct.
> 
> Can we check with DTC about Binbin's email, just to get confidence that the tmp_dir option is not necessary? It seems that they lean towards not needing it, but they might need to know what Binbin's email had said so everyone can know whether this is actually addressed by METplus.
> 
> Finally, if you haven't done it yet, please be sure you have fully tested numbers 2 and 3 of your list of tests above in the instructions. I want to be sure that you see them working before I do any PR testing on them. When they are tested, please let me know.
> 
> Thanks!
> 
> @AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness.

Alright, I just looped DTC into Binbin's response regarding tmp_dir. Also, I already tested 2 and 3, albeit 2 is out of scope for this PR, and I since reverted those modifications earlier this week upon discussion with Alicia and Shelley.

PerryShafran-NOAA commented:
@StevenSimon-NOAA Oh, OK.  We'll wait for DTC's answer before we make any changes to test number 1.

I didn't realize that test number 2 was the portion that was deemed out of scope for this PR.  

I'll give the ecme snowfall stats job a test and see how that goes.  Stay tuned.

PerryShafran-NOAA commented:
@StevenSimon-NOAA OK, here's the plan:

1) When we get the tmp_dir stuff sorted out, and if changes are needed, I will re-run Test 1 from above with any changes.  Then I will run them *again* using the beta3 versions of MET and METplus, to ensure that this fix is good for both versions.

2) I will also test the ecme snowfall job, using both versions as well.  

These tests are necessary because these fixes will be going into both the bugfix version (which uses the current production versions of met/metplus) and EVSv2.0 (which will use a later version of EVS, currently the beta3s are being used in the main parallel.  

Stay tuned for all.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA OK, here's the plan:
> 
> 1. When we get the tmp_dir stuff sorted out, and if changes are needed, I will re-run Test 1 from above with any changes.  Then I will run them _again_ using the beta3 versions of MET and METplus, to ensure that this fix is good for both versions.
> 2. I will also test the ecme snowfall job, using both versions as well.
> 
> These tests are necessary because these fixes will be going into both the bugfix version (which uses the current production versions of met/metplus) and EVSv2.0 (which will use a later version of EVS, currently the beta3s are being used in the main parallel.
> 
> Stay tuned for all.

DTC just replied to me regarding Binbin's reply. They are of the understanding that METplus 5.0.1 supports setting the tmp_dir based on the value set in the METplus config, which should default to a directory other than /tmp. It would only be set to /tmp if you explicitly set it to that value in the METplus config, which is unlikely if that is not permitted.

StevenSimon-NOAA committed:
Merge branch 'develop' into feature/global_ens_metplus_climo_config_bugfix. Add check for corrupt IMD files and check for forcast hour list length.

StevenSimon-NOAA commented:
Quick update from DTC. They confirmed that we can remove tmp_dir from the overrides option, since MetPlus 5.0.1 does properly handle TMP_DIR as it is set in these config files. I will make those updates.

StevenSimon-NOAA committed:
Remove tmp_dir from metplus config overrides option.

PerryShafran-NOAA commented:
OK great!  So here is the next step:

I will repeat the test that I just did earlier on with this PR, with the production versions of MET/METplus.  I will be using as input the production prep directories.  Then for the 2nd test, I will use the beta3 versions of MET/METplus using the emc.vpppg prep directories.  We do not need to compare the two tests against each other; we just need to make sure that this code works well with both the bugfix and the EVSv2 set ups.

First, with the production versions.  I'll let you know when that's done.

@AliciaBentley-NOAA @ShelleyMelchior-NOAA for awareness.

StevenSimon-NOAA committed:
Merge branch 'develop' into feature/global_ens_metplus_climo_config_bugfix. Implement Narre restart.

PerryShafran-NOAA commented:
@StevenSimon-NOAA Apologies, I started these jobs late Friday and then my workday ended.  Please check these job outputs:

The .o files are here:
 /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr467test/EVS/dev/drivers/scripts/stats/global_ens
 
 The output is here (look in the 20240417 directories):
 /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/global_ens 
 
 In this case, we should look to see that the stats file sizes are approximate, now knowing that they will different but not all having one larger than the other.  Compare these to production:  /lfs/h1/ops/prod/com/evs/v1.0/stats/global_ens.

StevenSimon-NOAA commented:
> @StevenSimon-NOAA Apologies, I started these jobs late Friday and then my workday ended. Please check these job outputs:
> 
> The .o files are here: /lfs/h2/emc/vpppg/noscrub/perry.shafran/pr467test/EVS/dev/drivers/scripts/stats/global_ens
> 
> The output is here (look in the 20240417 directories): /lfs/h2/emc/vpppg/noscrub/perry.shafran/evs/v1.0/stats/global_ens
> 
> In this case, we should look to see that the stats file sizes are approximate, now knowing that they will different but not all having one larger than the other. Compare these to production: /lfs/h1/ops/prod/com/evs/v1.0/stats/global_ens.

I identified no visceral ERRORs or WARNINGs in the log files. The log files indicate that the metplus_config_overrides indicate that the duplcate flag and obs_summary options were correctly processed. The small stats file counts are as expected. Upon comparison, the test atmos g2o large stats file sizes differ from their emc.vpppg counterparts for 20240417. Like last time, they are a mixed bag. But a quick look at the atmos g2o test large stats files reveals that they contain fewer observations and/or matched pairs for a given variable and metric, which confirms DTC's findings. Furthermore, I found that the test ecme snowfall large stats file matches its emc.vpppg counterpart both in size and content.

AliciaBentley-NOAA commented:
Hi @StevenSimon-NOAA. Thanks for doing this comparison. Can you please compare the test results to ops EVS v1.0 output as well (like @PerryShafran-NOAA mentioned at the bottom of his previous comment on Github)? I suspect you'll get the same results, but we really need to be comparing this test to ops EVS output and not to the emc.vpppg paralllel. Thank you!

PerryShafran-NOAA commented:
@StevenSimon-NOAA OK, I'm going to call this a win here.  I think your bottom line seems to be everything looks OK, correct me if I'm wrong.

Now what I'm going to do is to conduct the snowfall test using the prod versions of met/metplus, to compare with the production stats files.  Once we work through this one, I'll be conducting both tests again using the beta versions of met/metplus, to compare with emc.vpppg.  

Onto the snowfall test.  This one should be relatively quick.

StevenSimon-NOAA commented:
> Hi @StevenSimon-NOAA. Thanks for doing this comparison. Can you please compare the test results to ops EVS v1.0 output as well (like @PerryShafran-NOAA mentioned at the bottom of his previous comment on Github)? I suspect you'll get the same results, but we really need to be comparing this test to ops EVS output and not to the emc.vpppg paralllel. Thank you!

Same results when the test atmos g2o stats are compared to the ops prod atmos g2o stats. In each case, the number of matched pairs and/observations are reduced .

StevenSimon-NOAA commented:
> @StevenSimon-NOAA OK, I'm going to call this a win here. I think your bottom line seems to be everything looks OK, correct me if I'm wrong.
> 
> Now what I'm going to do is to conduct the snowfall test using the prod versions of met/metplus, to compare with the production stats files. Once we work through this one, I'll be conducting both tests again using the beta versions of met/metplus, to compare with emc.vpppg.
> 
> Onto the snowfall test. This one should be relatively quick.

Yep, the number of observations and/or matched pairs are reduced.

PerryShafran-NOAA commented:
@StevenSimon-NOAA Referring to the snowfall run?  You've already checked it out?

StevenSimon-NOAA commented:
> @StevenSimon-NOAA Referring to the snowfall run? You've already checked it out?

Yes, I did.

PerryShafran-NOAA commented:
@StevenSimon-NOAA OK, in that case, I will move on to the 2nd portion of the test, the use of the beta3 test versions (MET v12.0.0-beta3 and METplus v6.0.0-beta3).  I will start by running the first test (the override flag).  Please stay tuned for the results of that.

PerryShafran-NOAA commented:
@StevenSimon-NOAA Also, the next tests will be using the emc.vpppg prep data as input, and should be compared to the emc.vpppg output.  You do not need to look in the production directories for this one.

PerryShafran-NOAA commented:
@StevenSimon-NOAA All the runs, both g2o and the ecme snowfall, are complete.  Would you take a look at them (same directories as previous) and let me know that what you see is OK?  Let me know that you are now OK merging these changes.

StevenSimon-NOAA commented:
I think you forgot to run the ecme snowfall stats driver for the newest runs.

PerryShafran-NOAA commented:
You're right!  I thought I did.  Re-submitting it now.  Please check that when it's finished, as well as the other g2o runs.  I'm running to a few meetings now, let me know how things look.

StevenSimon-NOAA commented:
I identified no visceral or unexpected ERRORs or WARNINGs in the log files. There was a set of WARNINGs pertaining to deprecated CLIMO INPUT DIR and CLIMO INPUT TEMPLATE options, but those will be handled in a different PR down the road. The log files indicate that the metplus_config_overrides that the duplicate flag and obs_summary options were correctly processed. The small stats file counts are as expected in number. Upon comparison, the test atmos g2o large stats file sizes differ from their emc.vpppg counterparts for 20240420. Like last time, they are a mixed bag. But a quick look at the atmos g2o test large stats files reveals that they contain fewer observations and/or matched pairs for a given variable and metric, which confirms DTC's findings. Furthermore, I found that the test 20240421 ecme snowfall large stats file matches its emc.vpppg counterpart both in size and content.

PerryShafran-NOAA commented:
Excellent!  Looks like we're good here.  One more item, please do a sync fork to bring in changes from develop, and then we can merge this PR.

StevenSimon-NOAA committed:
Merge branch 'develop' into feature/global_ens_metplus_climo_config_bugfix. Implement Bugfix/cam severe prep.

StevenSimon-NOAA commented:
> Excellent! Looks like we're good here. One more item, please do a sync fork to bring in changes from develop, and then we can merge this PR.

Ok, the feature branch should now be up-to-date.

PerryShafran-NOAA commented:
Thank you!  Getting ready to merge.

PerryShafran-NOAA reviewed (approved):
Code works as expected.  Approved for merge.
